services/offline/src/lib/dhis2-connection-status.tsx

```diff
@@ -0,0 +1,201 @@
+            // if value changed, reset ping backoff to initial
+            if (newIsConnected !== prevIsConnected) {
+                smartIntervalRef.current?.resetDelayToInitial()
+            }
+            // if disconnected and EITHER 1. coming from connected or
+            // 2. there is no last-connect val, update the val in localStorage
+            if (!newIsConnected && (prevIsConnected || !getLastConnected())) {
+                updateLastConnected()
+            }
+            return newIsConnected
+        })
+    }, [])
+
+    // Note that the SW is configured to not cache ping requests and won't
+    // trigger `handleChange` below to avoid redundant signals. This also
+    // helps to detect the connectivity status when the SW is not available
+    // for some reason (maybe private browsing, first installation, or
+    // insecure browser context)
+    const ping = useCallback(() => {
+        return engine
+            .query(pingQuery)
+            .catch((err) => {
+                if (/Unexpected token 'p'/.test(err.message)) {
+                    // The request succeeded; this is just a weird endpoint
+                    // (rest of the error is '"pong" is not valid JSON')
+                    // todo: maybe change link handling of this endpoint
+                    return
+                } else {
+                    // It's a different error; throw to the next catch handler
+                    throw err
+                }
+            })
+            .then(() => {
+                // Ping is successful; set 'connected'
+                updateConnectedState(true)
+            })
+            .catch((err) => {
+                // Can get here if unauthorized, network error, etc.
+                console.error('Ping failed:', err.message)
+
+                // Only network errors should change status
+                const isNetworkErr = /^An unknown network error occurred$/.test(
+                    err.message
+                )
+                if (isNetworkErr) {
+                    updateConnectedState(false)
+                }
+            })
+    }, [engine, updateConnectedState])
+
+    /** Called when SW reports updates from incidental network traffic */
+    const onUpdate = useCallback(
+        ({ isConnected: newIsConnected }) => {
+            console.log('handling update from sw')
+            updateConnectedState(newIsConnected)
+            // Snooze ping timer to reduce pings since we know state from SW
+            smartIntervalRef.current?.snooze()
+        },
+        [updateConnectedState]
+    )
+
+    useEffect(() => {
+        const smartInterval = new SmartInterval({
+            initialDelay: 5000,
+            // don't ping if window isn't focused or visible
+            initialPauseValue:
+                !document.hasFocus() || document.visibilityState !== 'visible',
+            callback: ping,
+        })
+        smartIntervalRef.current = smartInterval
```

Comment:
Instead of creating a `SmartInterval` instance here and storing it in a ref, you could initialise it at the start of the function body. For that to make sense, the `SmartInterval` class should have an `init` method which you could call here and things wouldn't start ticking until you do.

Reply:
Added: [link to commit](https://github.com/dhis2/app-runtime/pull/1203/commits/f10365c51276de9733a5a17860963717cf14120c)

Reply:
I ended up going back on this because it would create a new SmartInterval instance on each render